### ISSUE TITLE
Check for failures when plugins are removed

### ISSUE THREAD

**Issue by corneliusweig** [kubernetes-sigs/krew#421]
Currently, it is not entirely clear, how `krew` will behave in the following circumstances:

* A user installed a plugin, and its manifest gets removed from `krew-index`
* A user installed a plugin, and its platform is removed from the corresponding manifest.

This issue is to explore if `krew` runs into any errors with one of its sub-commands in one of the above situations.

---

These scenarios can be tested by manipulating the `index` folder in `$KREW_ROOT` and setting the `--no-update-index` CLI option.

/good-first-issue

**Comment by ahmetb** [comment]
We currently print a really weird "Upgraded" message in case of 

> * A user installed a plugin, and its platform is removed from the corresponding manifest.

For example, I removed `darwin` platform of `ctx` plugin and ran:

````
$ kubectl krew upgrade ctx --no-update-index  -v=10
I1203 15:16:51.666127   66174 upgrade.go:81] --no-update-index specified, skipping updating local copy of plugin index
I1203 15:16:51.666209   66174 scanner.go:82] Reading plugin "ctx"
I1203 15:16:51.667079   66174 upgrade.go:65] Upgrading plugin: ctx
I1203 15:16:51.667286   66174 platform.go:43] Matching platform for labels(arch=amd64,os=darwin)
Upgraded plugin: ctx
```

It definitely didn't do an upgrade, but prints a message saying so.
/kind bug

> * A user installed a plugin, and its manifest gets removed from `krew-index`

I think this scenario is pretty straightforward. We treat it like the plugin was installed manually (`--manifest=.., --archive=..`) and wasn't part of the index to begin with.

When you do an `upgrade` on such a plugin (that is removed from index) today, you'll see:

```
F1203 15:19:40.001383   66283 root.go:58] failed to load the plugin manifest for plugin ns: open /Users/ahmetb/.krew/index/plugins/ns.yaml: no such file or directory
```

...which isn't nice, but can we do any better than this? I think this is OK. (Maybe we can treat errors.Is(os.IsNotExist) and say "plugin no longer exists in the index repository"?)

**Comment by chriskim06** [comment]
I had a few questions about this issue:

When I remove linux from the grep plugin I get this:
```
$ KREW_ROOT="$PWD/playground" out/bin/krew-linux_amd64 upgrade grep --no-update-index
F1207 16:32:40.568661   11046 root.go:58] failed to upgrade plugin "grep": plugin "grep" does not offer installation for this platform (linux/amd64)
```
instead of a message saying grep was updated. It seems like this got fixed in https://github.com/kubernetes-sigs/krew/pull/427.

> A user installed a plugin, and its platform is removed from the corresponding manifest.

This seems to have been taken care of?

Is this issue dependent on https://github.com/kubernetes-sigs/krew/issues/414 if we want to use `errors.Is(err, os.ErrNotExist)`?

It looks like `github.com/pkg/errors` is currently being used to wrap errors, so in order to use the go 1.13 `errors.Is` function, this would have to either use a named import (I think that's what its called) for `errors`/`github.com/pkg/errors` or remove the current package and wrap errors using the `%w` verb.

If that's all true is there a preference for one or the other?

**Comment by chriskim06** [comment]
I'd like to help out with this issue.

/assign

**Comment by corneliusweig** [comment]
Hey @chriskim06,

> This seems to have been taken care of?

Yes, this case is indeed covered already.

What we are looking for in this issue is to go through all of the subcommands and see if they behave in an unexpected way. The most critical ones are definitely `upgrade`/`update` and these have been fixed already.

---

If you want to improve the error messages for the case where a plugin manifest can't be found, you can do so with the go 1.13 `errors.Is` tooling. However, it's not dependent on #414, so you could already do this.

For the message, I lean more towards `plugin no longer exists in the index repository`, because we should focus on our majority of users who don't tamper with the `index` directory. For them, this error really only occurs, if the manifest is _no longer_ available.

Feel free to also open an issue for improving exactly this error message :)

**Comment by chriskim06** [comment]
Sounds good, thanks @corneliusweig, so basically we want to try removing the plugin/just the platform from the plugin for an installed plugin and try running through the subcommands for the plugin. I can do that and reply back here if I find anything weird.

I mentioned the error message because of @ahmetb's comment that the output could be a little more user friendly. I agree though that maybe that should be separate from the work involved with this. I would be happy to create a new issue for that later today!

**Comment by chriskim06** [comment]
Here's what I've seen so far with the other sub commands. I ran this locally by building the binary and messing with the local index as suggested in this issue (removing linux platform for `$KREW_ROOT/index/plugins/grep.yaml` and then later on deleting it altogether after installing it). I limited this to the sub commands that deal with plugins.

Removing the platform from the plugin manifest of an installed plugin:
`info`:
```
$ KREW_ROOT="$PWD/playground" out/bin/krew-linux_amd64 info grep
NAME: grep
URI: https://github.com/guessi/kubectl-grep/releases/download/v1.2.1/kubectl-grep-Linux-x86_64.tar.gz
SHA256: 2781447e3406ca8f90d49e1871f579c412836c9abe03ade8ae5f7c1d8cedd474
VERSION: v1.2.1
HOMEPAGE: https://github.com/guessi/kubectl-grep
DESCRIPTION:
Filter Kubernetes resources by matching their names

Examples:

List all pods in all namespaces
$ kubectl grep pods --all-namespaces

List all pods in namespace "star-lab" which contain the keyword "flash"
$ kubectl grep pods -n star-lab flash

No more pipe, built-in grep :-)

CAVEATS:
\
 | This plugin requires an existing KUBECONFIG file, with a `current-context` field set.
 |
 | Usage:
 |
 |   $ kubectl grep # output help messages
 |
 | More Info:
 | - https://github.com/guessi/kubectl-grep
/
```
`install`:
```
$ KREW_ROOT="$PWD/playground" out/bin/krew-linux_amd64 install grep --no-update-index
Installing plugin: grep
W1208 22:59:07.825033   12168 install.go:124] Skipping plugin "grep", it is already installed
```
`upgrade`:
```
$ KREW_ROOT="$PWD/playground" out/bin/krew-linux_amd64 upgrade grep --no-update-index
F1208 22:42:35.155907    6431 root.go:58] failed to upgrade plugin "grep": plugin "grep" does not offer installation for this platform (linux/amd64)
```
`uninstall`
```
$ KREW_ROOT="$PWD/playground" out/bin/krew-linux_amd64 uninstall grep
Uninstalled plugin grep
```
`search`:
```
$ KREW_ROOT="$PWD/playground" out/bin/krew-linux_amd64 search grep
NAME  DESCRIPTION                                         INSTALLED
grep  Filter Kubernetes resources by matching their n...  yes
```

Removing the plugin manifest:
`info`:
```
$ KREW_ROOT="$PWD/playground" out/bin/krew-linux_amd64 info grep
F1208 22:46:20.717002    7708 root.go:58] plugin "grep" not found
```
`install`:
```
$ KREW_ROOT="$PWD/playground" out/bin/krew-linux_amd64 install grep --no-update-index
F1208 22:47:09.143103    7974 root.go:58] failed to load plugin "grep" from the index: open /home/kimc/go/src/sigs.k8s.io/krew/playground/index/plugins/grep.yaml: no such file or directory
```
`upgrade`:
```
$ KREW_ROOT="$PWD/playground" out/bin/krew-linux_amd64 upgrade grep --no-update-index
F1208 22:47:56.737799    8298 root.go:58] failed to load the plugin manifest for plugin grep: open /home/kimc/go/src/sigs.k8s.io/krew/playground/index/plugins/grep.yaml: no such file or directory
```
`uninstall`
```
$ KREW_ROOT="$PWD/playground" out/bin/krew-linux_amd64 uninstall grep
Uninstalled plugin grep
```
`search`:
```
$ KREW_ROOT="$PWD/playground" out/bin/krew-linux_amd64 search grep

```

Everything seems ok to me. @corneliusweig does this seem alright?

**Comment by ahmetb** [comment]
Thanks for this thorough investigation @chriskim06, really appreciated. Looks ok. But you're right: Particular messages around "not found" plugins can be confusing.

Overall I'd treat this as a P3 bug for now. Some things that I see can be fixed:

- **krew install WRONG_NAME** (plugin not installed at all). Currently we print `failed to load plugin "shit" from the index: open /Users/ahmetb/.krew/index/plugins/wrong_name.yaml: no such file or directory` . &larr; Ugh, confusing. We can do `os.IsNotExist` error check (which is guaranteed from indexscanner.LoadPluginByName) and print a friendlier message like:

    > plugin WRONG_NAME does not exist in the plugin index.

- **krew install NAME** (already installed, but **manifest deleted**): according to @chriskim06 investigation, we try to load manifest from fs again, I wonder why. we should be treating it like `ErrAlreadyInstalled` and say it's already installed. (we don't have to fix this, again low priority, and can be fixed like previous item). 

- **krew upgrade NAME** (but manifest is deleted): Again, this can be fixed by relying on `os.IsNotExist` and `indexscanner.LoadPluginByName` (being added in #350) to print a friendlier message.

**Comment by chriskim06** [comment]
Ok, I'll create a new issue to try and capture what's been said here regarding the messaging around some of these scenarios.

**Comment by corneliusweig** [comment]
@chriskim06 Thanks for your thorough investigation. I'll close this in favor of https://github.com/kubernetes-sigs/krew/issues/431.

/close
